src/mlpack/methods/boostmetric/boostmetric_impl.hpp

```diff
@@ -0,0 +1,179 @@
+/**
+ * @file boostmetric_impl.hpp
+ * @author Manish Kumar
+ *
+ * Implementation of BoostMetric class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_BOOSTMETRIC_BOOSTMETRIC_IMPL_HPP
+#define MLPACK_METHODS_BOOSTMETRIC_BOOSTMETRIC_IMPL_HPP
+
+// In case it was not already included.
+#include "boostmetric.hpp"
+
+namespace mlpack {
+namespace boostmetric {
+
+/**
+ * Takes in a reference to the dataset. Copies the data, initializes
+ * all of the member variables and constraint object.
+ */
+template<typename MetricType>
+BOOSTMETRIC<MetricType>::BOOSTMETRIC(const arma::mat& dataset,
+                                     const arma::Row<size_t>& labels,
+                                     const size_t k) :
+    dataset(dataset),
+    labels(labels),
+    k(k),
+    maxIter(500),
+    tolerance(1e-7),
+    wTolerance(1e-5),
+    constraint(dataset, labels, k)
+{ /* nothing to do */ }
+
+// Calculate inner product of two matrices.
+inline double innerProduct(arma::mat& Ar, arma::mat& Z)
+{
+  double sum = 0.0;
+
+  for (size_t i = 0; i < Z.n_elem; i++)
+    sum += Ar(i) * Z(i);
+
+  return sum;
+}
+
+template<typename MetricType>
+void BOOSTMETRIC<MetricType>::LearnDistance(arma::mat& outputMatrix)
+{
+  Timer::Start("boostmetric");
+
+  // See if we were passed an initialized matrix.
+  if ((outputMatrix.n_rows != dataset.n_rows) ||
+      (outputMatrix.n_cols != dataset.n_rows))
+  {
+    outputMatrix.eye(dataset.n_rows, dataset.n_rows);
+  }
+
+  // Compute triplets.
+  arma::Mat<size_t> triplets;
+  constraint.Triplets(triplets, dataset, labels);
+
+  size_t N = triplets.n_cols;
+  size_t dim = dataset.n_rows;
+
+  // Initialize u.
+  arma::vec u;
+  u.ones(N);
+
+  // Normalize u such that sum(u_i) = 1
+  u = u / N;
+
+  arma::cube A(dim, dim, N);
+
+  // Initialize A (coefficient matrix of difference between distance
+  // of impostors & target neighbors).
+  for (size_t i = 0; i < N; i++)
+  {
+    A.slice(i) = (dataset.col(triplets(0, i)) - dataset.col(triplets(2, i))) *
+        arma::trans(dataset.col(triplets(0, i)) - dataset.col(triplets(2, i)))
+        -
+        (dataset.col(triplets(0, i)) - dataset.col(triplets(1, i))) *
+        arma::trans(dataset.col(triplets(0, i)) - dataset.col(triplets(1, i)));
+  }
+
+  // Declare weighted sum of Ar matrices.
+  arma::mat Acap;
+
+  // Now iterate!
+  for (size_t i = 0; i < maxIter; i++)
+  {
+    // Initialize Acap.
+    Acap.zeros(dim, dim);
+    for (size_t j = 0; j < N; j++)
+      Acap += u(j) * A.slice(j);
+
+    arma::vec eigval;
+    arma::mat eigvec;
+    arma::eig_sym(eigval, eigvec, (Acap + trans(Acap) / 2));
```

Comment:
I'd suggest timing this specific step also.  If we need we can modify Armadillo to use ARPACK to only get the largest eigenvalue, which can be a lot faster.

Reply:
I think it doesn't require any change, as we will be needing eigvector as well in very next calculation.
`arma::mat Z = eigvec.col(maxIndex(0)) * arma::trans(eigvec.col(maxIndex(0)));`

Reply:
It'll also take much longer for datasets with large dimensionality, so I'd be interested to see how long it takes for covertype (55d) and MNIST (784d).